### DISCUSSION TITLE
Harfbuzz installed without `--build-from-source` causes error during SuperTux deployment (CPack)

### DISCUSSION THREAD

**Discussion by tobbi** [Homebrew/discussions#4735] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.4
ORIGIN: https://github.com/Homebrew/brew
HEAD: 98f1be77f93a0d23fbc68629d0854522e9f0c1c3
Last commit: 2 weeks ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 0125edb525a6789b461e926d277445e8ce34c8d4
Core tap last commit: 11 days ago
Core tap branch: master
Core tap JSON: 12 Aug 19:44 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: ["--no-quarantine"]
HOMEBREW_COLOR: set
HOMEBREW_MAKE_JOBS: 3
HOMEBREW_NO_AUTO_UPDATE: set
HOMEBREW_NO_INSTALL_CLEANUP: set
Homebrew Ruby: 2.6.10 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: 3-core 64-bit ivybridge
Clang: 13.0.0 build 1300
Git: 2.41.0 => /usr/local/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 11.7.8-x86_64
CLT: 13.2.0.0.1.1638488800
Xcode: 13.2.1 => /Applications/Xcode_13.2.1.app/Contents/Developer
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  ruby@2.7

Warning: You have an unnecessary local Cask tap.
This can cause problems installing up-to-date casks.
Please remove it by running:
  brew untap homebrew/cask

Warning: You have an unnecessary local Core tap!
This can cause problems installing up-to-date formulae.
Please remove it by running:
 brew untap homebrew/core
```

### Description of issue

Please see https://github.com/actions/runner-images/issues/8078 for the history of this.

We're using CPack to package our executable for MacOS. Now, when we run cpack -G, we keep getting the following error:
```
CPack: Create package using Bundle
CPack: Install projects
CPack: - Run preinstall target for: SUPERTUX
CPack: - Install project: SUPERTUX []
CMake Error at /usr/local/Cellar/cmake/3.27.1/share/cmake/Modules/BundleUtilities.cmake:458 (message):
  otool -l failed: 1

  error:
  /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/otool-classic:
  can't open file: @loader_path/../../../../opt/glib/lib/libglib-2.0.0.dylib
  (No such file or directory)

Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.27.1/share/cmake/Modules/BundleUtilities.cmake:527 (get_item_rpaths)
  /usr/local/Cellar/cmake/3.27.1/share/cmake/Modules/BundleUtilities.cmake:649 (set_bundle_key_values)
  /usr/local/Cellar/cmake/3.27.1/share/cmake/Modules/BundleUtilities.cmake:933 (get_bundle_keys)
  /Users/runner/work/supertux/supertux/build/cmake_install.cmake:123 (fixup_bundle)


CPack Error: Error when generating package: SuperTux
usage: mv [-f | -i | -n] [-v] source target
       mv [-f | -i | -n] [-v] source ... directory
Error: Process completed with exit code 64.
```

Apparently, the error only occurs since the update to harfbuzz 8.1.1 and only when building without `--from-source`. Any ideas?

**Comment by SMillerDev** [comment]
Seems like an issue that the cpack devs should be able to help you with since this is not really a supported usecase for Homebrew.

**Reply by tobbi** [reply]
Shouldn't the versions installed with `--from-source` and without be functionally identical?

**Reply by SMillerDev** [reply]
They should be, but we stopped supporting source builds because it's really hard to know and adapt to all the ways that a user's system can be different from CI that would change the build.

**Comment by ZhongRuoyu** [comment]
For this problem specifically: `homebrew/core` bottles are built [with `HOMEBREW_RELOCATABLE_INSTALL_NAMES` set](https://github.com/Homebrew/homebrew-core/pull/137230), so relative install names are used (see https://github.com/Homebrew/brew/pull/15571). That's likely the reason why your local build behaved differently.

**Comment by vpolikarpov-akvelon** [comment]
Could this issue have emerged as a result of 4.1 release of Homebrew?

**Reply by SMillerDev** [reply]
In combinatie with the fact that this is not supported or tested, yes. It really could be

**Reply by vpolikarpov-akvelon** [reply]
I'm talking about installation without `--build-from-source` flag. The original problem emerged couple of weeks ago: builds started to fail with no obvious cause. I supposed first that failures were caused by harfbuzz update as it's rollback fixed the issue. But then @tobbi found that the issue fixes even when installing new version of harfbuzz, but only when building it from source.

Long story short:
- it was ok month ago with harfbuzz from homebrew without building from source
- it started failing couple of weeks ago (with harfbuzz from homebrew without building from source)
- it works good now only with harfbuzz installed with `--build-from-source`
- it fails now with harfbuzz installed with `--build-from-source` and `HOMEBREW_RELOCATABLE_INSTALL_NAMES` set, btw

So the question is: why harfbuzz brebuilt by homebrew stopped working recently?

**Reply by SMillerDev** [reply]
Nowhere does it say Harfbuzz stopped working though. The thread is about CPack failing to repack things.

**Reply by ah666677** [reply]
@SMillerDev

**Reply by ah666677** [reply]
‌ت

**Comment by ah666677** [comment]
> ### Output of `brew config`
> 
> ```shell
> HOMEBREW_VERSION: 4.1.4
> ORIGIN: https://github.com/Homebrew/brew
> HEAD: 98f1be77f93a0d23fbc68629d0854522e9f0c1c3
> Last commit: 2 weeks ago
> Core tap origin: https://github.com/Homebrew/homebrew-core
> Core tap HEAD: 0125edb525a6789b461e926d277445e8ce34c8d4
> Core tap last commit: 11 days ago
> Core tap branch: master
> Core tap JSON: 12 Aug 19:44 UTC
> HOMEBREW_PREFIX: /usr/local
> HOMEBREW_CASK_OPTS: ["--no-quarantine"]
> HOMEBREW_COLOR: set
> HOMEBREW_MAKE_JOBS: 3
> HOMEBREW_NO_AUTO_UPDATE: set
> HOMEBREW_NO_INSTALL_CLEANUP: set
> Homebrew Ruby: 2.6.10 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
> CPU: 3-core 64-bit ivybridge
> Clang: 13.0.0 build 1300
> Git: 2.41.0 => /usr/local/bin/git
> Curl: 7.87.0 => /usr/bin/curl
> macOS: 11.7.8-x86_64
> CLT: 13.2.0.0.1.1638488800
> Xcode: 13.2.1 => /Applications/Xcode_13.2.1.app/Contents/Developer
> ```
> 
> 
> ### Output of `brew doctor`
> 
> ```shell
> Please note that these warnings are just used to help the Homebrew maintainers
> with debugging if you file an issue. If everything you use Homebrew for is
> working fine: please don't worry or file an issue; just ignore this. Thanks!
> 
> Warning: Some installed formulae are deprecated or disabled.
> You should find replacements for the following formulae:
>   ruby@2.7
> 
> Warning: You have an unnecessary local Cask tap.
> This can cause problems installing up-to-date casks.
> Please remove it by running:
>   brew untap homebrew/cask
> 
> Warning: You have an unnecessary local Core tap!
> This can cause problems installing up-to-date formulae.
> Please remove it by running:
>  brew untap homebrew/core
> ```
> 
> 
> ### Description of issue
> 
> Please see https://github.com/actions/runner-images/issues/8078 for the history of this.
> 
> We're using CPack to package our executable for MacOS. Now, when we run cpack -G, we keep getting the following error:
> ```
> CPack: Create package using Bundle
> CPack: Install projects
> CPack: - Run preinstall target for: SUPERTUX
> CPack: - Install project: SUPERTUX []
> CMake Error at /usr/local/Cellar/cmake/3.27.1/share/cmake/Modules/BundleUtilities.cmake:458 (message):
>   otool -l failed: 1
> 
>   error:
>   /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/otool-classic:
>   can't open file: @loader_path/../../../../opt/glib/lib/libglib-2.0.0.dylib
>   (No such file or directory)
> 
> Call Stack (most recent call first):
>   /usr/local/Cellar/cmake/3.27.1/share/cmake/Modules/BundleUtilities.cmake:527 (get_item_rpaths)
>   /usr/local/Cellar/cmake/3.27.1/share/cmake/Modules/BundleUtilities.cmake:649 (set_bundle_key_values)
>   /usr/local/Cellar/cmake/3.27.1/share/cmake/Modules/BundleUtilities.cmake:933 (get_bundle_keys)
>   /Users/runner/work/supertux/supertux/build/cmake_install.cmake:123 (fixup_bundle)
> 
> 
> CPack Error: Error when generating package: SuperTux
> usage: mv [-f | -i | -n] [-v] source target
>        mv [-f | -i | -n] [-v] source ... directory
> Error: Process completed with exit code 64.
> ```
> 
> Apparently, the error only occurs since the update to harfbuzz 8.1.1 and only when building without `--from-source`. Any ideas?

h

**Comment by ZhongRuoyu** [comment, marked as answer]
FYI: this should be addressed. See Homebrew/homebrew-core#140930 for more information.

**Reply by tobbi** [reply]
Thank you so much! I will try to update our CI build files

**Reply by ah666677** [reply]
tz

**Comment by tobbi** [comment]
@ZhongRuoyu A very similar issue re-emerged with a different dependency yesterday:
```
CMake Error at /usr/local/Cellar/cmake/3.28.1/share/cmake/Modules/BundleUtilities.cmake:458 (message):
  otool -l failed: 1

  error:
  /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/otool-classic:
  can't open file: @rpath/libjxl_cms.0.9.dylib (No such file or directory)

Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.28.1/share/cmake/Modules/BundleUtilities.cmake:527 (get_item_rpaths)
  /usr/local/Cellar/cmake/3.28.1/share/cmake/Modules/BundleUtilities.cmake:649 (set_bundle_key_values)
  /usr/local/Cellar/cmake/3.28.1/share/cmake/Modules/BundleUtilities.cmake:933 (get_bundle_keys)
  /Users/runner/work/supertux/supertux/build/cmake_install.cmake:123 (fixup_bundle)


CPack Error: Error when generating package: SuperTux
usage: mv [-f | -i | -n] [-v] source target
       mv [-f | -i | -n] [-v] source ... directory
```

I don't know enough about the bug reporting process or the right steps to take, but I'm sure you can help with rebottling libjxml.

What does "rebottling" even do?

**Reply by ZhongRuoyu** [reply]
@tobbi Sorry, my inbox was overflowing and I did not have enough bandwidth. Will take a look at this soon.

"Rebottling" means to rebuild a formula / binary package. The original problem in this discussion was solved by rebuilding certain libraries, because their dynamic dependencies' install name contained `@loader_path` which, albeit valid, CMake's `fixup_bundle` could not recognise.

**Reply by ZhongRuoyu** [reply]
SuperTux/supertux#2779 should hopefully fix that. No rebottling needed -- the issue is with `@rpath` (similar to `libwebp`'s; see https://github.com/SuperTux/supertux/pull/2622#issuecomment-1706816598), so [this existing workaround](https://github.com/SuperTux/supertux/blob/9b4eaeed047c2b202b4449a540832a5361b32cf5/.github/workflows/macos.yml#L57-L60) can be reused.

**Reply by tobbi** [reply]
Thanks! I tried working around it but obviously I had no idea what I was doing. How do you find the correct paths to use for the command?

**Reply by ZhongRuoyu** [reply]
You need to identify which dylib caused this issue. The error occurred while opening `@rpath/libjxl_cms.0.9.dylib`, so it must come from something that depends on it, and has `libjxl_cms.0.9.dylib`'s parent directory as an `rpath` entry. That usually is another dylib from the same package (`jpeg-xl`). Looking at the other dylibs from the same package:

<details>
<summary> <code>find "$(brew --prefix jpeg-xl)/lib" -name '*.dylib' -type f -exec otool -L '{}' \;</code> </summary>

```console
$ find "$(brew --prefix jpeg-xl)/lib" -name '*.dylib' -type f -exec otool -L '{}' \;
/usr/local/opt/jpeg-xl/lib/libjxl_threads.0.9.1.dylib:
	/usr/local/opt/jpeg-xl/lib/libjxl_threads.0.9.dylib (compatibility version 0.9.0, current version 0.9.1)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1600.157.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)
/usr/local/opt/jpeg-xl/lib/libjxl_cms.0.9.1.dylib:
	/usr/local/opt/jpeg-xl/lib/libjxl_cms.0.9.dylib (compatibility version 0.9.0, current version 0.9.1)
	/usr/local/opt/little-cms2/lib/liblcms2.2.dylib (compatibility version 3.0.0, current version 3.16.0)
	/usr/local/opt/highway/lib/libhwy.1.dylib (compatibility version 1.0.0, current version 1.0.7)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1600.157.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)
/usr/local/opt/jpeg-xl/lib/libjxl.0.9.1.dylib:
	/usr/local/opt/jpeg-xl/lib/libjxl.0.9.dylib (compatibility version 0.9.0, current version 0.9.1)
	@rpath/libjxl_cms.0.9.dylib (compatibility version 0.9.0, current version 0.9.1)
	/usr/local/opt/highway/lib/libhwy.1.dylib (compatibility version 1.0.0, current version 1.0.7)
	/usr/local/opt/brotli/lib/libbrotlidec.1.dylib (compatibility version 1.0.0, current version 1.1.0)
	/usr/local/opt/brotli/lib/libbrotlicommon.1.dylib (compatibility version 1.0.0, current version 1.1.0)
	/usr/local/opt/brotli/lib/libbrotlienc.1.dylib (compatibility version 1.0.0, current version 1.1.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1600.157.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)
/usr/local/opt/jpeg-xl/lib/libjxl_extras_codec.0.9.1.dylib:
	/usr/local/opt/jpeg-xl/lib/libjxl_extras_codec.0.9.dylib (compatibility version 0.9.0, current version 0.9.1)
	/usr/local/opt/giflib/lib/libgif.dylib (compatibility version 0.0.0, current version 7.2.0)
	/usr/local/opt/jpeg-turbo/lib/libjpeg.8.dylib (compatibility version 8.0.0, current version 8.3.2)
	/usr/local/opt/libpng/lib/libpng16.16.dylib (compatibility version 57.0.0, current version 57.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.12)
	@rpath/libjxl.0.9.dylib (compatibility version 0.9.0, current version 0.9.1)
	/usr/local/opt/openexr/lib/libOpenEXR-3_2.31.dylib (compatibility version 31.0.0, current version 31.3.2)
	/usr/local/opt/openexr/lib/libOpenEXRUtil-3_2.31.dylib (compatibility version 31.0.0, current version 31.3.2)
	/usr/local/opt/openexr/lib/libOpenEXRCore-3_2.31.dylib (compatibility version 31.0.0, current version 31.3.2)
	/usr/local/opt/openexr/lib/libIex-3_2.31.dylib (compatibility version 31.0.0, current version 31.3.2)
	/usr/local/opt/openexr/lib/libIlmThread-3_2.31.dylib (compatibility version 31.0.0, current version 31.3.2)
	/usr/local/opt/imath/lib/libImath-3_1.29.dylib (compatibility version 29.0.0, current version 29.8.0)
	@rpath/libjxl_cms.0.9.dylib (compatibility version 0.9.0, current version 0.9.1)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1600.157.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)
```
</details>

As shown in the output, `libjxl` and `libjxl_extras_codec` both have an `@rpath/libjxl_cms.0.9.dylib` entry. I tried changing the dylib install name from `libjxl` first (i.e. changing `@rpath/libjxl_cms.0.9.dylib` to a full path as in other entries; check [`man install_name_tool`](https://www.unix.com/man-page/osx/1/install_name_tool/)):

```
install_name_tool -change '@rpath/libjxl_cms.0.9.dylib' /usr/local/opt/jpeg-xl/lib/libjxl_cms.0.9.dylib /usr/local/opt/jpeg-xl/lib/libjxl.dylib
```

And that turned out to have fixed it.
